go/vt/vtctl/grpcvtctldserver/query.go

```diff
@@ -0,0 +1,300 @@
+			return nil, err
+		}
+	}
+
+	sm.TabletFailure, err = row[21].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.Progress, err = row[22].ToFloat32()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.MigrationContext = row[23].ToString()
+	// ddl_action row[24]
+	sm.Message = row[25].ToString()
+
+	sm.EtaSeconds, err = row[26].ToInt64()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.RowsCopied, err = row[27].ToUint64()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.TableRows, err = row[28].ToInt64()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.AddedUniqueKeys, err = row[29].ToUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.RemovedUniqueKeys, err = row[30].ToUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.LogFile = row[31].ToString()
+
+	sm.ArtifactRetention, err = valueToVTDuration(row[32], "s")
+	if err != nil {
+		return nil, err
+	}
+
+	sm.PostponeCompletion, err = row[33].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.RemovedUniqueKeyNames = row[34].ToString()
+	sm.DroppedNoDefaultColumnNames = row[35].ToString()
+	sm.ExpandedColumnNames = row[36].ToString()
+	sm.RevertibleNotes = row[37].ToString()
+
+	sm.AllowConcurrent, err = row[38].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.RevertedUuid = row[39].ToString()
+
+	sm.IsView, err = row[40].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.ReadyToComplete, err = row[41].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.VitessLivenessIndicator, err = row[42].ToInt64()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.UserThrottleRatio, err = row[43].ToFloat32()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.SpecialPlan = row[44].ToString()
+
+	sm.LastThrottledAt, err = valueToVTTime(row[45])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.ComponentThrottled = row[46].ToString()
+
+	sm.CancelledAt, err = valueToVTTime(row[47])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.PostponeLaunch, err = row[48].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.Stage = row[49].ToString()
+
+	sm.CutoverAttempts, err = row[50].ToUint32()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.IsImmediateOperation, err = row[51].ToBool()
+	if err != nil {
+		return nil, err
+	}
+
+	sm.ReviewedAt, err = valueToVTTime(row[52])
+	if err != nil {
+		return nil, err
+	}
+
+	sm.ReadyToCompleteAt, err = valueToVTTime(row[53])
+	if err != nil {
+		return nil, err
+	}
+
+	return sm, nil
+}
+
+func valueToVTTime(value sqltypes.Value) (*vttime.Time, error) {
+	s := value.ToString()
+	if s == "" {
+		return nil, nil
+	}
+
+	gotime, err := time.ParseInLocation(mysqlTimestampLayout, s, time.Local)
+	if err != nil {
+		return nil, err
+	}
+
+	return protoutil.TimeToProto(gotime), nil
+}
+
+func valueToVTDuration(value sqltypes.Value, defaultUnit string) (*vttime.Duration, error) {
+	s := value.ToString()
+	if s == "" {
+		return nil, nil
+	}
+
+	switch s[len(s)-1] {
+	case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
+		s += defaultUnit
```

Comment:
this could use some actual code commentary, but, essentially: "if the string ends with a digit, it's missing a duration unit, so stick on our default"